### PR TITLE
Migrate parent POM file and annotate critical methods

### PR DESCRIPTION
**Pull request by oleg-nenashev**
- [x] Migrate parent pom to 2.x
- [x] Make the tests compilable
- [x] Fix Javadoc critical issues
- [x] Annotate methods, which may really return null
- [x] Fix the tests to work with new HtmlUnit
- [x] Fix critical issues discovered by FindBugs (NPEs, etc.)

Issues being addressed:
* https://issues.jenkins-ci.org/browse/JENKINS-34187
* https://issues.jenkins-ci.org/browse/JENKINS-32237
* https://issues.jenkins-ci.org/browse/JENKINS-19940
* https://issues.jenkins-ci.org/browse/JENKINS-22632

Disclaimer: I know about javax.annotation. They already exist in the plugin, hence I'll perform a bulk change later.

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
retriggering the build

**Comment by oleg-nenashev** [comment]
Ready for @reviewbybees in any case.
CC @fbelzunc

**Comment by fbelzunc** [comment]
```
ERROR: Maven JVM terminated unexpectedly with exit code 137
```

**Comment by oleg-nenashev** [comment]
Not sure what's going on with CI. Works on my laptop.

**Comment by oleg-nenashev** [comment]
rebuilding with JDK8

**Comment by fbelzunc** [comment]
I think it might be related for the type of slaves we are using to build. Are they the fastest and biggest ones?

**Comment by oleg-nenashev** [comment]
> I think it might be related for the type of slaves we are using to build. Are they the fastest and biggest ones?

They we m1.large, switched to hi-speed

**Comment by oleg-nenashev** [comment]
yay, hi-speed worked. JobDSL tests hammer the infra a bit

File: src/main/java/hudson/plugins/promoted_builds/FakeParent.java [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import java.io.IOException;
 import java.util.Collection;
 
+//TODO: needs polishing
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/promoted_builds/FakeParent.java]: Postponed it as it is not related to the task

File: pom.xml [hunk 2]
```diff
@@ -138,6 +133,30 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.apache.httpcomponents</groupId>
+      <artifactId>httpcore</artifactId>
+      <version>4.4.4</version>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 
   <build>
```
Comment (armfergom) [inline pom.xml]: The build section can be removed completely.  `maven-release-plugin`, `findbugs-maven-plugin` and `maven-hpi-plugin` are already defined in the parent pom.

File: pom.xml [hunk 3]
```diff
@@ -138,6 +133,30 @@
       <version>2.30</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
Comment (armfergom) [inline pom.xml]: Why are these new dependencies now required? (`matrix-auth` and `matrix-project`)

File: pom.xml [hunk 4]
```diff
@@ -15,6 +15,7 @@
   <url>http://wiki.jenkins-ci.org/display/JENKINS/Promoted+Builds+Plugin</url>
 
   <properties>
+    <jenkins.version>1.565</jenkins.version>
     <findbugs-maven-plugin.version>3.0.3</findbugs-maven-plugin.version>
     <findbugs.failOnError>true</findbugs.failOnError>
```
Comment (armfergom) [inline pom.xml]: These two properties (`findbugs.failOnError`and `findbugs-maven-plugin.version`) can go away when the build section is removed.

**Comment by armfergom** [comment]
@oleg-nenashev had some comments. All related to the changes in the pom file. Apart from that, it all LGTM.

File: pom.xml (hunk 3 above)
Comment (oleg-nenashev) [inline pom.xml]: They are being used in the test suites

Comment (armfergom) [inline pom.xml]: What was puzzling me was I was not seeing any **new** references to classes of those projects but the dependencies are new. They were being brought from the 1.565 parent pom I presume, sorry!

**Comment by armfergom** [comment]
Thanks @oleg-nenashev , 🐝 from me :)

**Comment by oleg-nenashev** [comment]
The test crashed

**Comment by armfergom** [comment]
With which error? I cloned your branch and performed a `mvn clean install` before :bee:ing the PR. Build was successful.

**Comment by oleg-nenashev** [comment]
I think it's an INFRA issue. Not a blocker

**Comment by oleg-nenashev** [comment]
BTW, I'm aware that there was no such issue before changing java.level

**Comment by armfergom** [comment]
Just some more information, `java.level` is used by:

* `maven-enforcer-plugin`
* `maven-compiler-plugin` 
* `animal-sniffer-maven-plugin`

Looking into the issue now, but it does seem like an infra issue as you pointed out

**Comment by Sauraus** [comment]
Would it be too much to ask for the PR to be squashed before final merge?

**Comment by oleg-nenashev** [comment]
@Sauraus I don't think squash is really required here, but I'll do it

**Comment by oleg-nenashev** [comment]
@Sauraus done

**Comment by Sauraus** [comment]
Thank you.

**Pull request merged**